File: c++/leetcode/108. Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  TreeNode *sortedArrayToBST(vector<int> &nums)
  {
    if (nums.empty())
      return nullptr;

    auto begin = nums.begin();
    auto end = nums.end();
    auto mid = nums.size() / 2;

    auto root = new TreeNode(nums[mid]);

    if (nums.size() == 1)
      return root;

    vector<int> left(begin, begin + mid);
    vector<int> right(vector<int>(begin + (mid + 1), end));

    root->left = sortedArrayToBST(left);
    root->right = sortedArrayToBST(right);

    return root;
  }
};

File: javascript/leetcode/108. Convert Sorted Array to Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function (nums) {
  if (!nums.length) return null;

  const mid = Math.floor(nums.length / 2);
  const root = new TreeNode(nums[mid]);

  if (nums.length === 1) return root;

  let right = nums.splice(mid + 1);
  let left = nums.splice(0, mid);

  root.left = sortedArrayToBST(left);
  root.right = sortedArrayToBST(right);

  return root;
};
